include/hipSYCL/sycl/libkernel/generic/hiplike/atomic_builtins.hpp

```diff
@@ -517,11 +556,18 @@ __acpp_atomic_fetch_min(double *addr, double x, memory_order order,
 template <access::address_space S, class T>
 HIPSYCL_HIPLIKE_BUILTIN T __acpp_atomic_fetch_max(
     T *addr, T x, memory_order order, memory_scope scope) noexcept {
+#if defined(__HIP__)
+  // HIP does not support some of the overloads of atomicMax until ROCm 5.7
+  // https://github.com/ROCm/clr/issues/2
+  return __hip_atomic_fetch_max(addr, x, builtin_memory_order(order),
+      __HIP_MEMORY_SCOPE_AGENT);
```

Comment:
`__HIP_MEMORY_SCOPE_AGENT` is a pessimization for all scoped that are weaker than `device`, and incorrect in case a user wants to use `system` memory scope.
You will probably need a switch statement, like SSCP does: https://github.com/AdaptiveCpp/AdaptiveCpp/blob/7677cf6eefd8ab46d66168cd07ab042109448124/src/libkernel/sscp/amdgpu/atomic.cpp#L51

Reply:
It should be the same as before, as `atomicMax` has device scope, and I didn't want to widen the scope (pun intended) of the PR. In any case I'm blocked by the SSCP issue and have no way of investigating this further 🤷‍♂️

Reply:
Ah, well... SMCP atomics are not great. I forgot that they don't handle that :( Alright then.

It might be useful to add some diagnostics to figure out which test exactly fails (which data type and operation). The atomic tests currently seem to be implemented as many checks in one test, so we can't see clearly where exactly things go wrong. Maybe it's not min/max with long/unsigned long that fails.

One difference I see in the atomic implementation between SMCP and SSCP is that SMCP uses some explicit CAS loops, whereas SSCP always calls the builtin and lets LLVM handle putting in CAS loops where needed.

Once I am back at the office I can take a look too.

PS: Happy holidays :)

Reply:
> Maybe it's not min/max with long/unsigned long that fails.

You were right! I've added a context message to print the data type and operation, and it looks like ROCm SSCP is [failing for `atomicMin` and `atomicMax` on `float` and `double`](https://github.com/AdaptiveCpp/AdaptiveCpp/actions/runs/12545162134/job/34978935338?pr=1666#step:10:343). Since [I changed](https://github.com/AdaptiveCpp/AdaptiveCpp/pull/1666/files#diff-e1eca73484f076818fe8dd3928b360ceadd5bc3c1d9e66a2fbd38986d2315962R84) the way the input data is initialized from `i` to `-i` (with `i` in `[0, 2047]`), the results are clearly wrong. In fact, it looks like the two operations are doing exactly the opposite..?!

> PS: Happy holidays :)

Thanks! And to you!